linen_examples/pixelcnn/sample.py

```diff
@@ -0,0 +1,126 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Sampling from PixelCNN++ using fixed-point iteration.
+"""
+from functools import partial
+
+from absl import app
+from absl import flags
+
+import numpy as onp
+from PIL import Image
+
+import jax
+from jax import random
+import jax.numpy as jnp
+
+import pixelcnn
+import train
+
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_integer(
+    'sample_batch_size', default=64,
+    help=('Batch size for sampling.'))
+
+flags.DEFINE_integer(
+    'sample_rng_seed', default=0,
+    help=('Random number generator seed for sampling.'))
+
+def generate_sample(pcnn_module, batch_size, rng_seed=0):
+  rng = random.PRNGKey(rng_seed)
+  rng, model_rng = random.split(rng)
+
+  # Create a model with dummy parameters and a dummy optimizer
+  example_images = jnp.zeros((1, 32, 32, 3))
+  model = train.create_model(model_rng, example_images, pcnn_module)
+  optimizer = train.create_optimizer(model, 0)
+
+  # Load learned parameters
+  _, ema = train.restore_checkpoint(optimizer, model.params)
+  model = model.replace(params=ema)
+
+  # Initialize batch of images
+  device_count = jax.local_device_count()
+  assert not batch_size % device_count, (
+      'Sampling batch size must be a multiple of the device count, got '
+      'sample_batch_size={}, device_count={}.'.format(batch_size,
+                                                      device_count))
+  sample_prev = jnp.zeros((device_count, batch_size // device_count, 32, 32, 3))
+
+  # and batch of rng keys
+  sample_rng = random.split(rng, device_count)
+
+  # Generate sample using fixed-point iteration
+  sample = sample_iteration(sample_rng, model, sample_prev)
+  while jnp.any(sample != sample_prev):
+    sample_prev, sample = sample, sample_iteration(sample_rng, model, sample)
+  return jnp.reshape(sample, (batch_size, 32, 32, 3))
+
+def _categorical_onehot(rng, logit_probs):
+  """Sample from a categorical distribution and one-hot encode the sample.
+  """
+  nr_mix = logit_probs.shape[-3]
+  idxs = random.categorical(rng, logit_probs, axis=-3)
+  return jnp.moveaxis(idxs[..., jnp.newaxis] == jnp.arange(nr_mix), -1, -3)
+
+def conditional_params_to_sample(rng, conditional_params):
+  means, inv_scales, logit_probs = conditional_params
+  rng_mix, rng_logistic = random.split(rng)
+  # Add channel dimension to one-hot mixture indicator
+  mix_indicator = _categorical_onehot(rng_mix, logit_probs)[..., jnp.newaxis]
+  # Use the mixture indicator to select the mean and inverse scale
+  mean      = jnp.sum(means      * mix_indicator, -4)
+  inv_scale = jnp.sum(inv_scales * mix_indicator, -4)
+  sample = mean + random.logistic(rng_logistic, mean.shape) / inv_scale
+  return snap_to_grid(sample)
+
+@partial(jax.pmap, static_broadcasted_argnums=1)
```

Comment:
> This means you will recompile for each set of parameters.

That doesn't appear to be happening, I wonder if that's because the static_broadcasted args are hashed on their `id`, and I don't think they're treated as pytrees, and the model is stateful/mutable? I could be wrong, this is just a guess :).

Reply:
Ah sorry my bad, I think `sample.py` hasn't been tested and you're probably right that it's broken as you describe (I thought that the comment was on `train.py`).

Reply:
Actually, I was wrong both times: because this is in the sampling code, not training, it _is_ ok to treat the model as static.

Reply:
Ok so assuming Jamie is correct I'm resolving this :-).